fix(controller): apply HideSecretData to server-side diff results for Secrets (#25260)

downfa11 opened the pull request:
<!--
Note on DCO:

If the DCO action in the integration test fails, one or more of your commits are not signed off. Please click on the *Details* link next to the DCO action for instructions on how to resolve this.
-->

Fixes #25260

## Summary  
  
Fixes Secret resources showing persistent OutOfSync status when using `IncludeMutationWebhook=true` with mutation webhooks.  
  
## Problem  
  
Applies `hideSecretData` normalization to desired state(Git) but not to server-side diff dry-run results, causing false OutOfSync status for Secrets modified by mutation webhooks.  

## Solution  
  
Wraps the KubeApplier returned by `getServerSideDiffDryRunApplier` to apply consistent `HideSecretData` normalization to both desired and live states.  
  
<br>

Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [x] The title of the PR conforms to the [Title of the PR](https://argo-cd.readthedocs.io/en/latest/developer-guide/submit-your-pr/#title-of-the-pr)
* [x] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [ ] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [x] Does this PR require documentation updates?
* [ ] I've updated documentation as required by this PR.
* [x] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/blob/master/community/CONTRIBUTING.md#legal)
* [x] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [x] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)).
* [ ] My new feature complies with the [feature status](https://github.com/argoproj/argoproj/blob/master/community/feature-status.md) guidelines.
* [x] I have added a brief description of why this PR is necessary and/or what this PR solves.
* [ ] Optional. My organization is added to USERS.md.
* [ ] Optional. For bug fixes, I've indicated what older releases this fix should be cherry-picked into (this may or may not happen depending on risk/complexity).

<!-- Please see [Contribution FAQs](https://argo-cd.readthedocs.io/en/latest/developer-guide/faq/) if you have questions about your pull-request. -->

downfa11 committed:
fix: normalize secrets in server-side diff

Signed-off-by: downfa11 <downfa11@naver.com>

downfa11 committed:
refactor: resolved lint issues

Signed-off-by: downfa11 <downfa11@naver.com>

downfa11 committed:
refactor: lint

Signed-off-by: downfa11 <downfa11@naver.com>

downfa11 committed:
refactor: resolved lint issues in sync_test

Signed-off-by: downfa11 <downfa11@naver.com>

ppapapetrou76 commented:
@downfa11 can you please resolve the conflicts - I will review the PR then

ppapapetrou76 commented on controller/sync_test.go:
```diff
@@ -1653,3 +1661,136 @@ func dig(obj any, path ...any) any {
 	return i
 }
+
+func TestSecretNormalizingApplier(t *testing.T) {
```
The unit test is good, but it should also cover:
- The case where inner.ApplyResource returns a non-JSON string (ensure the wrapper does not return an error).
- The case where the result is for a non-core group Secret or other resource (ensure only core/v1 Secret is normalized).

ppapapetrou76 commented on controller/sync.go:
```diff
@@ -83,7 +87,47 @@ func (m *appStateManager) getServerSideDiffDryRunApplier(cluster *v1alpha1.Clust
+}
+
+// secretNormalizingApplier wraps a KubeApplier to normalize Secret data
+type secretNormalizingApplier struct {
```
I think you should also implement the `ConvertToVersion` function, something like

```go
func (s *secretNormalizingApplier) ConvertToVersion(obj *unstructured.Unstructured, gv, targetGV string) (*unstructured.Unstructured, error) {
    return s.inner.ConvertToVersion(obj, gv, targetGV)
}
```

downfa11 committed:
fix: Improve secret normalization handling

Signed-off-by: downfa11 <downfa11@naver.com>

downfa11 committed:
fix: resolved lint issues

Signed-off-by: downfa11 <downfa11@naver.com>

ppapapetrou76 reviewed (approved):
LGTM 🏅  - thanks for addressing my comments

Vovcharaa commented:
Is there an ETA for merging this?

1ovsss commented:
Hey @alexmt @crenshaw-dev @jannfis  -- sorry for the ping, I know you weren't involved in this PR.

This fix has been approved and ready to merge since January 2026 but is blocked on approver permissions. @ppapapetrou76 suggested pinging @argoproj/argocd-approvers for a review.

We're hitting this bug in production -- ServerSideDiff=true with a mutation webhook (bank-vaults) causes all Secrets to show false OutOfSync due to hideSecretData normalization not being applied to server-side diff results. The only workaround is reverting to ignoreDifferences for all Secrets, which defeats the purpose of ServerSideDiff for secret change detection.

Could you take a look when you get a chance? Even if there are changes needed, any feedback would help move this forward. Thanks!

downfa11 committed:
chore: resolved minor issues

Signed-off-by: downfa11 <downfa11@naver.com>

downfa11 committed:
chore: resolved issues

Signed-off-by: downfa11 <downfa11@naver.com>

downfa11 committed:
chore: resolved go.mod issues

Signed-off-by: downfa11 <downfa11@naver.com>

downfa11 committed:
chore: go format

Signed-off-by: downfa11 <downfa11@naver.com>

pjiang-dev reviewed (approved):
LGTM

agaudreault commented on controller/sync.go:
```diff
@@ -78,7 +82,57 @@ func (m *appStateManager) getServerSideDiffDryRunApplier(cluster *v1alpha1.Clust
+	}
+
+	// Apply hideSecretData
+	if resultObj.GetKind() == kube.SecretKind && (resultObj.GroupVersionKind().Group == "" || resultObj.GroupVersionKind().Group == "core") {
```
Other places in the code apply the logic to `if res.Kind == kube.SecretKind && res.Group == ""`. This should be consistent. 

If there is a bug with the "core" group, that should be fixed in another PR and everything should be updated for consistency

agaudreault commented on controller/sync.go:
```diff
@@ -78,7 +82,57 @@ func (m *appStateManager) getServerSideDiffDryRunApplier(cluster *v1alpha1.Clust
+
+	wrappedOps := &secretNormalizingApplier{
+		inner:       ops,
+		settingsMgr: m.settingsMgr,
```
instead of saving the settingsManager that will fetch the ConfigMap on every call, save the sensitiveAnnotations in the secretNormalizingApplier struct. 

A new applier is created on each diff, so there is no need to always get the latest value from the CM.

agaudreault commented on controller/sync.go:
```diff
@@ -78,7 +82,57 @@ func (m *appStateManager) getServerSideDiffDryRunApplier(cluster *v1alpha1.Clust
+	}
+
+	isCoreSecret := obj.GetKind() == kube.SecretKind && (obj.GroupVersionKind().Group == "" || obj.GroupVersionKind().Group == "core")
+	if dryRunStrategy != cmdutil.DryRunServer || !isCoreSecret {
```
Shouldn't it be `dryRunStrategy == cmdutil.DryRunNone`? 

From my understanding, if this is not a dryRun, then we do not want to hide the secret data 🤔

downfa11 replied:
@agaudreault Thank you for the review! I have addressed the other comments. Regarding the `dryRunStrategy` condition, I would like to clarify the intent.

The goal is to hide secret data during the diff output for security while ensuring the actual data is used during real application. Since ArgoCD uses dry-run to generate diffs, the current logic masks secrets only when `dryRunStrategy == cmdutil.DryRunServer`.

If we were to use `dryRunStrategy == cmdutil.DryRunNone`, it would mask the data during actual application, which is the opposite of the intended behavior. 

The included unit tests verify that secrets are masked in the diff output but remain intact during actual application.

agaudreault replied:
@downfa11 Are you saying when we have `cmdutil.DryRunClient`, the output will be used to apply the resource later? 🤔 

My concern was that we should hide sensitive data whether it is DryRunClient or DryRunServer

agaudreault commented on controller/sync_test.go:
```diff
@@ -1768,3 +1775,180 @@ func TestValidateSyncPermissions(t *testing.T) {
+		_, gitHasPassword := gitData["password"]
+		_, resultHasPassword := resultData["password"]
+		assert.True(t, gitHasPassword, "git secret should have password field after normalization")
+		assert.True(t, resultHasPassword, "result secret should have password field after normalization")
```
This part of the test is not testing the secret normalization applier, but testing the logic of `HideSecretData` and `Diff`. It should be removed from the unit test

agaudreault commented on controller/sync_test.go:
```diff
@@ -1768,3 +1775,180 @@ func TestValidateSyncPermissions(t *testing.T) {
+			"metadata": map[string]any{
+				"name":      "test-secret",
+				"namespace": "default",
+			},
```
The unit test should validate that the sensitive annotations are also correctly masked.

leoluz reviewed (commented):
@pjiang-dev Besides Alex's comments, we need to make sure that this is not somehow overlapping/conflicting with Alex M. PR to fix the SSD bug.

pjiang-dev commented:
> @pjiang-dev Besides Alex's comments, we need to make sure that this is not somehow overlapping/conflicting with Alex M. PR to fix the SSD bug.

I double checked with the other SSD bug by Alex M. They are completely separate code paths as that one is the CLI/API only and this one is with the controller. There are no conflicts

downfa11 committed:
fix: resolved review issues

Signed-off-by: downfa11 <downfa11@naver.com>

downfa11 committed:
chore: resolved minor issues

Signed-off-by: downfa11 <downfa11@naver.com>

pjiang-dev reviewed (changes requested):
@downfa11 Looking at this again i have concerns that we should actually move this hiding secret to be in the diff logic instead of the Applier logic

The DryRunApplier's job is to ask the API server a question and return the answer.  Similar to normalization, masking secret makes sense be in diff logic instead of applier.

I feel a better place would be in serverSideDiff(), right after Normalize(), checking if the resource is a core Secret and calling HideSecretData on both predictedLive and live together.

leoluz commented on controller/sync.go:
```diff
@@ -78,7 +82,57 @@ func (m *appStateManager) getServerSideDiffDryRunApplier(cluster *v1alpha1.Clust
+}
+
+// secretNormalizingApplier wraps a KubeApplier to normalize Secret data
+type secretNormalizingApplier struct {
```
It seems that this is mixing 2 distinct responsibilities:
1. Dryrun apply resources
2. Normalize the dryrun result

I feel like we should keep the dryrunApplier with a single responsibility to execute the dryrun and the caller being the one with the responsibility to enforce the normalization logic.

pjiang-dev commented:
@downfa11 i hope you don't mind but since its been awhile and i would like to get this PR merged to make ServerSideDiff default, i addressed the comments and created a PR based off of your fork. (Your commits will be in that PR)

downfa11 commented:
@pjiang-dev Thanks for following up and for addressing the comments. 

Sorry for the delay on my side. I’m totally fine with you taking this forward and creating the PR from your side.

agaudreault commented:
Closing as this was merged in another PR.
@downfa11 You are listed as co-author of https://github.com/argoproj/argo-cd/commit/a7b70b5c8b13ce5bdabbe6b9ec61f8686439bc36.
Thank you both for fixing that.
